_episodes/05b-advanced-containers.md

```diff
@@ -0,0 +1,315 @@
+- Docker allows containers to read and write files from the Docker host.
+---
+
+In order to create and use your own containers, you may need more information than
+our previous example. You may want to use files from outside the container, copy
+those files into the container, and just generally learn a little bit about software
+installation. This episode will cover these. Note that the examples will get gradually
+more and more complex - most day-to-day use of containers can be accomplished
+using the first 1-2 sections on this page.
+
+## Using scripts and files from outside the container
+
+In your shell, change to the `sum` folder in the `docker-intro` folder.
+~~~
+$ cd ~/Desktop/docker-intro/sum
+~~~
+{: .language-bash}
+
+This folder has both a `Dockerfile` and a python script called `sum.py`. Let's say
+we wanted to try running the script using our recently created `alpine-python`
+container.
+
+> ## Running containers
+>
+> What command would we use to run python from the `alpine-python` container?
+{: .challenge}
+
+If we try running the container and Python script, what happens?
+~~~
+$ docker run alice/alpine-python python sum.py
+~~~
+{: .language-bash}
+~~~
+python: can't open file 'sum.py': [Errno 2] No such file or directory
+~~~
+{: .output}
+
+> ## No such file or directory
+>
+> What does the error message mean? Why might the Python inside the container
+> not be able to find or open our script?
+>
+{: challenge}
+
+The problem here is that the container and its file system is separate from our
+host computer's file system. When the container runs, it can't see anything outside
+itself, including any of the files on our computer. In order to use Python
+(inside the container) and our script (outside the container, on our computer),
+we need to create a link between the folder on our computer and the container.
+
+This link is called a "mount" and we can create it using an additional
+option to `docker run`. We'll also use the variable `$PWD` which will substitute
+in our current working directory. The option will look like this
+
+`-v $PWD:/temp`
+
+What this means is -- link my current folder with the container, and inside the
+container, name the folder `/temp`
+
+Let's try running the command now:
+~~~
+$ docker run -v $PWD:/temp alice/alpine-python python sum.py
+~~~
+{: .language-bash}
+
+But we get the same error!
+~~~
+python: can't open file 'sum.py': [Errno 2] No such file or directory
+~~~
+{: .output}
+
+This final piece is a bit tricky -- we really have to remember to put ourselves
+inside the container. Where is the `sum.py` file? It's in the folder that's been
+mapped to `/temp` -- so we need to include that in the path to the script. This
+command should give us what we need:
+~~~
+$ docker run -v $PWD:/temp alice/alpine-python python /temp/sum.py
+~~~
+{: .language-bash}
+
+> ## Exercise: Explore the script
+>
+> What happens if you use the `docker run` command above
```

Comment:
Can you add a solution here too?  Mostly because I have no idea what happens. 😆   Does the script take arguments?

Reply:
Yes. :) Should be obvious what happens (I hope)